motor/discovery/gcp/resolver_folder.go

```diff
@@ -0,0 +1,112 @@
+package gcp
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/errors"
+	"go.mondoo.com/cnquery/motor/asset"
+	"go.mondoo.com/cnquery/motor/discovery/common"
+	"go.mondoo.com/cnquery/motor/platform/detector"
+	"go.mondoo.com/cnquery/motor/providers"
+	gcp_provider "go.mondoo.com/cnquery/motor/providers/google"
+	"go.mondoo.com/cnquery/motor/providers/resolver"
+	"go.mondoo.com/cnquery/motor/vault"
+)
+
+type GcpFolderResolver struct{}
+
+func (k *GcpFolderResolver) Name() string {
+	return "GCP Folder Resolver"
+}
+
+func (r *GcpFolderResolver) AvailableDiscoveryTargets() []string {
+	return []string{
+		common.DiscoveryAuto, common.DiscoveryAll, DiscoveryProjects,
+		DiscoveryInstances, DiscoveryComputeImages, DiscoveryComputeNetworks, DiscoveryComputeSubnetworks, DiscoveryComputeFirewalls,
+		DiscoveryGkeClusters,
+		DiscoveryStorageBuckets,
+		DiscoveryBigQueryDatasets,
+	}
+}
+
+func (r *GcpFolderResolver) Resolve(ctx context.Context, tc *providers.Config, credsResolver vault.Resolver, sfn common.QuerySecretFn, userIdDetectors ...providers.PlatformIdDetector) ([]*asset.Asset, error) {
+	resolved := []*asset.Asset{}
+	if tc == nil || tc.Options["folder-id"] == "" {
+		return resolved, nil
+	}
+
+	// Note: we use the resolver instead of the direct gcp_provider.New to resolve credentials properly
+	m, err := resolver.NewMotorConnection(ctx, tc, credsResolver)
+	if err != nil {
+		return nil, err
+	}
+	defer m.Close()
+	provider, ok := m.Provider.(*gcp_provider.Provider)
+	if !ok {
+		return nil, errors.New("could not create gcp provider")
+	}
+
+	identifier, err := provider.Identifier()
+	if err != nil {
+		return nil, err
+	}
+
+	// detect platform info for the asset
+	detector := detector.New(provider)
+	pf, err := detector.Platform()
+	if err != nil {
+		return nil, err
+	}
+
+	folderId := tc.Options["folder-id"]
+	md, err := NewMQLAssetsDiscovery(provider)
+	if err != nil {
+		return nil, err
+	}
+
+	folder, err := GetValue[string](md, fmt.Sprintf("return gcp.folder(id: '%s').name", folderId))
+	if err != nil {
+		return nil, err
+	}
+
+	var resolvedRoot *asset.Asset
+	if tc.IncludesOneOfDiscoveryTarget(DiscoveryFolders) {
+		pf.Name = "gcp-folder"
+		resolvedRoot = &asset.Asset{
+			PlatformIds: []string{identifier},
+			Name:        "GCP folder " + folder,
+			Platform:    pf,
+			Connections: []*providers.Config{tc}, // pass-in the current config
+		}
+		resolved = append(resolved, resolvedRoot)
+	}
+
+	if tc.IncludesOneOfDiscoveryTarget(common.DiscoveryAuto, common.DiscoveryAll,
+		DiscoveryInstances, DiscoveryComputeImages, DiscoveryComputeNetworks, DiscoveryComputeSubnetworks, DiscoveryComputeFirewalls,
```

Comment:
I guess I'm missing something here. I do not understand why we try to resolve folders, when we discover, e.g. instances. My understanding is, that folders only contain projects, not resources directly.

Do you have an example, in which case this is needed?

Reply:
I can see why this is confusing. I will refactor the discovery flags a bit to make it more clear. Essentially, what is happening here is that a GCP folder contains projects. We, then re-use the project resolver to discover all assets under those projects.

Example:
you can run `cnquery scan gcp folder --folder-id 3421312 --discover compute-instance`. We need to go through this code, the project asset won't be discovered but via the project discovery we will detect the compute instances.

Reply:
Now I explicitly reuse the discovery targets between projects, folders and orgs. Just to make it obvious they are sort of nested

Reply:
Thanks for the explanation.